service-base/src/main/java/org/eclipse/hono/service/tenant/TenantHttpEndpoint.java

```diff
@@ -68,155 +69,104 @@ public String getName() {
-        final JsonObject requestMsg = TenantConstants.getServiceRequestAsJson(Action.ACTION_UPDATE.toString(),
-                tenantId, payload);
+        final JsonObject requestMsg = TenantConstants.getServiceRequestAsJson(action.toString(), tenantId, payload);
 
-        doTenantAction(ctx, requestMsg, (status, tenantResult) -> {
-                response.setStatusCode(status);
-                if (status >=400) {
+        sendTenantAction(ctx, requestMsg, (status, tenantResult) -> {
+            response.setStatusCode(status);
+            if (status >=400) {
+                setResponseBody(tenantResult, response);
+            } else if (sendResponseForStatus != null) {
+                if (sendResponseForStatus.test(status)) {
+                    if (action == Action.ACTION_ADD) {
+                        response.putHeader(
+                                HttpHeaders.LOCATION,
+                                String.format("/%s/%s", TenantConstants.TENANT_ENDPOINT, tenantId));
+                    }
                     setResponseBody(tenantResult, response);
                 }
-                response.end();
+            }
+            response.end();
         });
     }
 
-    private void doRemoveTenant(final RoutingContext ctx) {
+    private static String getTenantParam(final RoutingContext ctx) {
+        return ctx.request().getParam(PARAM_TENANT_ID);
+    }
 
-        final String tenantId = getTenantParam(ctx);
-        logger.debug("removing tenant [tenant: {}]", tenantId);
-        final HttpServerResponse response = ctx.response();
-        final JsonObject requestMsg = TenantConstants.getServiceRequestAsJson(Action.ACTION_REMOVE.toString(), tenantId);
-        doTenantAction(ctx, requestMsg, (status, tenantResult) -> {
-                response.setStatusCode(status);
-                if (status >= 400) {
-                    setResponseBody(tenantResult, response);
-                }
-                response.end();
-        });
+    private static String getTenantParamFromPayload(final JsonObject payload) {
+        return (payload != null ? (String) payload.remove(TenantConstants.FIELD_TENANT_ID) : null);
+    }
+
+    private static void setResponseBody(final JsonObject tenantResult, final HttpServerResponse response) {
+        final JsonObject msg = tenantResult.getJsonObject(TenantConstants.FIELD_PAYLOAD);
+        if (msg != null) {
+            final String body = msg.encodePrettily();
+            response.putHeader(HttpHeaders.CONTENT_TYPE, HttpUtils.CONTENT_TYPE_JSON_UFT8)
+                .putHeader(HttpHeaders.CONTENT_LENGTH, String.valueOf(body.length()))
+                .write(body);
+        }
     }
 
-    private void doTenantAction(final RoutingContext ctx, final JsonObject requestMsg, final BiConsumer<Integer, JsonObject> responseHandler) {
+    private void sendTenantAction(final RoutingContext ctx, final JsonObject requestMsg, final BiConsumer<Integer, JsonObject> responseHandler) {
```

Comment:
we should be able to pull this up into `AbstractHttpEndpoint`, don't you think? There doesn't seem to be anything left that is tenant specific, is there?